### PR TITLE
Multi_Setup_Trainer: let users to train a method with multiple setups, including different seeds and hdims

### PR DESCRIPTION
**Pull request by Ttopiac**
It is a framework for us to train self-play agents with different seeds and hdims simultaneously and allow us to resume the training if that fails. It is a framework. Following it, we can also implement the same thing for other methods, like n-1-sp and other methods.

### CODE REVIEW

File: oai_agents/agents/rl.py [hunk 1]
```diff
@@ -302,33 +312,43 @@ def train_agents(self, total_train_timesteps, tag_for_returning_agent, exp_name=
...
-                self.ck_list.append(({k: 0 for k in self.args.layout_names}, path, tag))
+                self.save_init_model_and_cklist()
+
 
         best_path, best_tag = None, None
 
         self.steps = self.start_step
-        curr_timesteps = self.start_timestep
-        prev_timesteps = self.learning_agent.num_timesteps
+        self.learning_agent.num_timesteps = self.n_envs*self.start_timestep
```
Comment (avv-va) [inline oai_agents/agents/rl.py:341]: Why do we need to change this?

File: oai_agents/agents/rl.py [hunk 2]
```diff
@@ -302,33 +312,43 @@ def train_agents(self, total_train_timesteps, tag_for_returning_agent, exp_name=
...
+        print(f"curr_timesteps: {self.learning_agent.num_timesteps}")
```
Comment (avv-va) [inline oai_agents/agents/rl.py]: Delete print

File: oai_agents/common/arguments.py [hunk 3]
```diff
@@ -83,6 +83,9 @@ def get_arguments(additional_args=[]):
     parser.add_argument("--num-dynamic-advs-per-heatmap", type=int, default=2)
     parser.add_argument("--custom-agent-ck-rate-generation", type=int)
 
+    parser.add_argument('--for-evaluation', type=bool, default=False, help="The trained agents are used for evaluating other agents. Please note that seeds and h_dim are different when agents are trained for evaluating others.)")
```
Comment (avv-va) [inline oai_agents/common/arguments.py]: Can you explain what 'for-evaluation' enables?

File: oai_agents/common/arguments.py [hunk 4]
```diff
@@ -83,6 +83,9 @@ def get_arguments(additional_args=[]):
...
+    parser.add_argument("--num-of-training-variants", type=int, default=4)
```
Comment (avv-va) [inline oai_agents/common/arguments.py]: Same for here

File: oai_agents/common/checked_model_name_handler.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        return bool(self.pattern.match(tag))
+
+    def get_all_checked_tags(self, path: Union[Path, None] = None) -> List[str]:
+        """
+        Retrieve all valid checked model tags (subdirectories) under the specified path that match the pattern.
+
+        :param path: The directory path to search for valid checked model tags. Can be a Path object or None.
+        :return: A list of valid checked model tag names.
+        :raises ValueError: If the path is None.
```
Comment (avv-va) [inline oai_agents/common/checked_model_name_handler.py]: You don't need to mention the potential errors it will raise in the function description

File: oai_agents/common/multi_setup_trainer.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        if self.num_of_training_variants <= len(seeds):
+            selected_seeds = seeds[:self.num_of_training_variants]
+            selected_hdims = hdims[:self.num_of_training_variants]
+        else:
+            selected_seeds = seeds[:]
+            selected_hdims = hdims[:]
+
+            remaining = self.num_of_training_variants - len(seeds)
+            available_seeds = list(set(range(min_seed, max_seed + 1)) - set(selected_seeds))
+            random_seeds = random.sample(available_seeds, remaining)
```
Comment (avv-va) [inline oai_agents/common/multi_setup_trainer.py]: why do we reuse a seed if there aren't enough seeds in training_seeds? Shouldn't we just make sure that training_seeds and evaluation_seeds always have enough unique seeds?

File: oai_agents/common/population.py [hunk 7]
```diff
@@ -147,16 +185,17 @@ def save_population(args, population):
...
-                   ck_rate,
-                   total_training_timesteps,
-                   train_types,
-                   eval_types,
-                   num_SPs_to_train,
-                   unseen_teammates_len=0,
-                   force_training=False,
-                   tag=KeyCheckpoints.MOST_RECENT_TRAINED_MODEL,
-                   ):
+def get_categorized_SP_population(
```
Comment (avv-va) [inline oai_agents/common/population.py]: What do you mean by 'categorized'?

File: oai_agents/common/teammates_collection.py [hunk 8]
```diff
@@ -267,24 +267,25 @@ def generate_TC(args,
 
 
 def get_best_SP_agent(args, population):
+    # all_agents = [agent for agent in population[args.layout_names[0]]]
```
Comment (avv-va) [inline oai_agents/common/teammates_collection.py]: delete

File: scripts/train_agents.py [hunk 9]
```diff
@@ -197,52 +232,64 @@ def N_1_FCP(args):
...
+        n_1_fcp_curriculum=n_1_fcp_curriculum,
+        unseen_teammates_len=unseen_teammates_len
+    )
 
 
 def set_input(args):
     args.num_players = args.teammates_len + 1
 
     two_chefs_layouts = [
+        'coordination_ring',
```
Comment (avv-va) [inline scripts/train_agents.py]: Could you please separate standard layouts from aamas layouts:

aamas layouts = [selected_ ..., ]
standard layouts = [...]

\# two_chefs_layouts = standard_layouts 
two_chefs_layouts = aamas_layouts

File: tests/test_oai_agents/test_base_agent.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
+from oai_agents.agents.base_agent import OAITrainer
```
Comment (avv-va) [inline tests/test_oai_agents/test_base_agent.py:1]: Nice!

File: oai_agents/agents/rl.py (hunk 1 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py:341]: If n_env is 10, and the start_timestep is 100, then the timestep we count in RLagentlearner will be 1000. Therefore, to measure the num_timesteps for learning_agent, this line of code is necessary.

I assume most of us will be still confused here. I can clarify it in our meeting or we can set up a zoom meeting to specifically discuss it.

File: oai_agents/common/arguments.py (hunk 3 above)
Comment (Ttopiac) [reply oai_agents/common/arguments.py]: I will update "help" there.

if for_evaluation is true, we generate population of agents for evaluation. This is currently only for population of self-play agents since we have not implemented them for other methods.

File: oai_agents/common/arguments.py (hunk 4 above)
Comment (Ttopiac) [reply oai_agents/common/arguments.py]: if num_of_training_variants is 4, we generate 4  agents. This is currently only for population of self-play agents since we have not implemented them for other methods.

File: oai_agents/common/multi_setup_trainer.py (hunk 6 above)
Comment (Ttopiac) [reply oai_agents/common/multi_setup_trainer.py]: We did not reuse seeds here. 

We firstly have 8 unique seeds for both training and evaluation. However, if we need more, we sample some other seeds from 0 to 2999 for training and some other from 3000 to 5999 for evaluation.

I am unsure whether I answered your concerns. If not, feel free to let me know. Thanks!

File: oai_agents/agents/rl.py (hunk 2 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py]: will do!

File: oai_agents/common/checked_model_name_handler.py (hunk 5 above)
Comment (Ttopiac) [reply oai_agents/common/checked_model_name_handler.py]: will delete them!

File: oai_agents/common/population.py (hunk 7 above)
Comment (Ttopiac) [reply oai_agents/common/population.py]: The function `get_population` was originally a clear name because it solely generated a population of agents. Now, its clarity is diminished because the code has been updated to support two functionalities: 

1. Generating a categorized population of agents (low, medium, and high performance).
2. Generating a population of uncategorized agents, supported by `Multi_Setup_Trainer`.

Since the latter functionality does not involve categorization, the name `get_population`, used by the first, may no longer fully represent the function's purpose.

File: scripts/train_agents.py (hunk 9 above)
Comment (Ttopiac) [reply scripts/train_agents.py]: Yes, I will define special layouts and standard layouts for that. 

But I assume we still train agents for both instead of setting up a separate training.

File: oai_agents/agents/il.py [hunk 11]
```diff
@@ -185,7 +185,7 @@ def run_epoch(self, agent_idx):
         self.agents[agent_idx].eval()
         return np.mean(losses)
 
-    def train_agents(self, epochs=100, exp_name=None):
+    def train_agents(self, epochs=100):
```
Comment (Jmiceli17) [inline oai_agents/agents/il.py:188]: since we're removing the `exp_name` argument, the reference to the argument on line 192 should be removed (just use `self.args.exp_name`

File: oai_agents/agents/rl.py [hunk 12]
```diff
@@ -260,19 +270,16 @@ def wrap_agent(self, sb3_agent, name):
...
-    def get_experiment_name(self, exp_name):
-        return exp_name or str(self.args.exp_dir) + '/' + self.name
-
 
     def should_evaluate(self, steps):
         mean_training_rew = np.mean([ep_info["r"] for ep_info in self.learning_agent.agent.ep_info_buffer])
         self.best_training_rew *= 1
 
-        steps_divisable_by_15 = (steps + 1) % 15 == 0
+        steps_divisible_by_x = (steps + 1) % 15 == 0
```
Comment (Jmiceli17) [inline oai_agents/agents/rl.py:278]: should we make `15` a config parameter?

File: oai_agents/agents/rl.py [hunk 13]
```diff
@@ -302,33 +312,42 @@ def train_agents(self, total_train_timesteps, tag_for_returning_agent, exp_name=
...
         best_path, best_tag = None, None
 
         self.steps = self.start_step
-        curr_timesteps = self.start_timestep
-        prev_timesteps = self.learning_agent.num_timesteps
+        self.learning_agent.num_timesteps = self.n_envs*self.start_timestep
+        ck_name_handler = CheckedModelNameHandler()
 
-        while curr_timesteps < total_train_timesteps:
+        while self.learning_agent.num_timesteps < total_train_timesteps:
```
Comment (Jmiceli17) [inline oai_agents/agents/rl.py:344]: To echo the comment here: https://github.com/HIRO-group/multiHRI/pull/55/files#r1887217671
Changing `curr_timesteps` to `self.learning_agent.num_timesteps` drastically changes this training loop because `curr_timesteps` was updated each iteration of the while loop.

File: oai_agents/agents/rl.py [hunk 14]
```diff
@@ -302,33 +312,42 @@ def train_agents(self, total_train_timesteps, tag_for_returning_agent, exp_name=
...
             self.curriculum.update(current_step=self.steps)
             self.set_new_teammates(curriculum=self.curriculum)
 
             # In each iteration the agent collects n_envs * n_steps experiences. This continues until self.learning_agent.num_timesteps > epoch_timesteps is reached.
             self.learning_agent.learn(self.epoch_timesteps)
-
-            curr_timesteps += self.learning_agent.num_timesteps - prev_timesteps
-            prev_timesteps = self.learning_agent.num_timesteps
+            self.steps += 1
```
Comment (Jmiceli17) [inline oai_agents/agents/rl.py:350]: why was this moved here?

File: oai_agents/agents/rl.py [hunk 15]
```diff
@@ -339,14 +358,20 @@ def train_agents(self, total_train_timesteps, tag_for_returning_agent, exp_name=
...
                     if self.learning_agent.num_timesteps // self.checkpoint_rate > (len(self.ck_list) - 1):
-                        path, tag = self.save_agents(tag=f'ck_{len(self.ck_list)}_rew_{mean_reward}')
+                        path = OAITrainer.get_model_path(
+                            base_dir=self.args.base_dir,
+                            exp_folder=self.args.exp_dir,
+                            model_name=self.name
+                        )
+                        tag = ck_name_handler.generate_tag(id=len(self.ck_list), mean_reward=mean_reward)
                         self.ck_list.append((rew_per_layout, path, tag))
+                        path, tag = self.save_agents(path=path, tag=tag)
```
Comment (Jmiceli17) [inline oai_agents/agents/rl.py]: it looks like `path` and `tag` are unused later in this function so I'd recommend replacing them with  `_, _`

File: oai_agents/common/arguments.py (hunk 4 above)
Comment (Jmiceli17) [reply oai_agents/common/arguments.py]: I would rename so it is more clear that we are talking about the total number of output agents. 
Maybe 
`total-ego-agents` or `total-main-agents` or `num-of-final-product-agents` etc.

File: oai_agents/common/multi_setup_trainer.py (hunk 6 above)
Comment (Jmiceli17) [reply oai_agents/common/multi_setup_trainer.py]: instead of randomly generating seeds for the extra agents does it make sense to throw an error and require the user to input additional seeds? This would require a slight modification to the constructor of `MultiSetupTrainer`

File: tests/test_oai_agents/test_base_agent.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+        print("Test passed: Tags returned as expected.")
+    else:
+        print(f"Test failed: Expected {expected_tags}, but got {checked_tags}")
+
+    # Clean up the test directories after the test
+    # This will remove the entire "agent_models/test_agents_folder" structure created for testing
+    shutil.rmtree(base_dir / "agent_models" / "test_agents_folder")
+
+# Run the test function
+test_list_agent_checked_tags()
```
Comment (Jmiceli17) [inline tests/test_oai_agents/test_base_agent.py]: Recommend putting this inside a `__main__` function

File: scripts/train_agents.py [hunk 17]
```diff
@@ -197,52 +232,62 @@ def N_1_FCP(args):
...
 
     three_chefs_layouts = [
         'selected_3_chefs_coordination_ring',
         'selected_3_chefs_counter_circuit',
         'selected_3_chefs_cramped_room',
-        # 'selected_3_chefs_double_counter_circuit',
-        # 'selected_3_chefs_secret_coordination_ring',
-        # 'selected_3_chefs_spacious_room_few_resources',
-        # 'selected_3_chefs_spacious_room_no_counter_space',
-        # 'selected_3_chefs_storage_room'
```
Comment (Jmiceli17) [inline scripts/train_agents.py:238]: These should be removed from 4 and 5 chef layouts as well if we remove them from the 3 chef layouts

File: oai_agents/common/population.py [hunk 18]
```diff
@@ -105,30 +102,71 @@ def ensure_we_will_have_enough_agents_in_population(teammates_len,
...
+    else:
+        # Use all predefined settings
+        selected_seeds = seeds[:]
+        selected_hdims = hdims[:]
 
+        # Generate additional random settings if more agents are needed
+        remaining = num_of_required_agents - len(seeds)
+        available_seeds = list(set(range(min_seed, max_seed)) - set(selected_seeds))
+        random_seeds = random.sample(available_seeds, remaining)  # Generate random seeds
+        random_hdims = [256] * remaining # Generate random hidden dimensions
```
Comment (Jmiceli17) [inline oai_agents/common/population.py]: same comment here as in https://github.com/HIRO-group/multiHRI/pull/55#discussion_r1887237248

File: oai_agents/common/population.py (hunk 7 above)
Comment (Jmiceli17) [reply oai_agents/common/population.py]: I'm confused by the name change as well. I don't see how this function is providing categorized agents. The role of [generate_TC()](https://github.com/HIRO-group/multiHRI/blob/main/oai_agents/common/teammates_collection.py#L186) is to separate the population into categories. Is that function unnecessary now?

File: oai_agents/common/teammates_collection.py [hunk 19]
```diff
@@ -267,24 +267,24 @@ def generate_TC(args,
...
-        all_agents = [agent for agent in population[layout_name]]
-
     for agent in all_agents:
         scores = [agent.layout_scores[layout_name] for layout_name in args.layout_names]
         agents_scores_averaged_over_layouts.append((agent, sum(scores)/len(scores)))
     best_agent = max(agents_scores_averaged_over_layouts, key=lambda x: x[1])
     return best_agent[0]
 
-
+def get_all_agents(layout_name, population):
```
Comment (Jmiceli17) [inline oai_agents/common/teammates_collection.py]: recommend renaming this to `get_all_agents_for_layout()`

File: oai_agents/common/arguments.py (hunk 4 above)
Comment (Ttopiac) [reply oai_agents/common/arguments.py]: total-ego-agents is good. I will rename it by this new name.

File: oai_agents/common/multi_setup_trainer.py (hunk 6 above)
Comment (Ttopiac) [reply oai_agents/common/multi_setup_trainer.py]: I thought about that. But if we let user to input additional seeds, that may not be easy to use. Do you have any suggestion changes for us to input additional seeds? (If we just let them to use arguments, I think that will make the code confused to understand. We may need other ways to implement.)

File: oai_agents/common/population.py (hunk 7 above)
Comment (Ttopiac) [reply oai_agents/common/population.py]: I read the function again. It looks like get_categorized_population is not a good name. However, get_population is not, either. We need other names for this one for us to not get confused when we have so many other population generation methods. Look for other good alternatives? Any suggestions will be appreicated!!

File: oai_agents/agents/rl.py (hunk 13 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py:344]: It was not updated anymore. 

It was redundant lines of code. We updated curr_timesteps using self.learning_agent.num_timesteps and prev_timesteps, but curr_timesteps was actually self.learning_agent.num_timesteps.

File: oai_agents/agents/rl.py (hunk 14 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py:350]: It was a bug.

Comment (Ttopiac) [reply oai_agents/agents/rl.py:350]: If we add step after the if condition, then we saved incorrect self.steps information. We fixed that bug by moving it to line 350

File: oai_agents/agents/rl.py (hunk 15 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py]: no problems! I will replace it.

File: oai_agents/agents/rl.py (hunk 14 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py:350]: This bug causes us to resume from incorrect checkpoints, which was usually a step before what it should be.

File: oai_agents/common/teammates_collection.py (hunk 19 above)
Comment (Ttopiac) [reply oai_agents/common/teammates_collection.py]: Good suggestion. I will rename it.

File: tests/test_oai_agents/test_base_agent.py (hunk 16 above)
Comment (Ttopiac) [reply tests/test_oai_agents/test_base_agent.py]: good recommendation! I will do it.

**Review by narendasan — DISMISSED** [review]
Ignore, for some reason only one commit rendered in the diff

File: oai_agents/common/arguments.py (hunk 3 above)
Comment (avv-va) [reply oai_agents/common/arguments.py]: I see! can you rename the argument to something more descriptive? for example: gen-pop-for-eval

File: oai_agents/common/multi_setup_trainer.py (hunk 6 above)
Comment (avv-va) [reply oai_agents/common/multi_setup_trainer.py]: I propose just hard coding enough seeds so we don't have to randomly sample new seed. This way, the seeds stay the same during each run, but if we randomly sample seeds, we won't get similar seed. 

I also agree with Joe, that we should throw an error if the user requires more seed. In that case, the user should just come and update the function with more seeds (it doesn't have to be an additional argument)

File: oai_agents/common/arguments.py (hunk 4 above)
Comment (avv-va) [reply oai_agents/common/arguments.py]: reminder to replace num_SPs_to_train with this var as well

**Comment by avv-va** [comment]
The code looks good! After your address the comments, please run the code in quick_test mode for 2,3 and 5 players using the SPN_XSPCKP and the default variables on the main branch for the quick test mode to make sure the functionality of the rest of the functions are not broken. Thanks!

File: oai_agents/common/multi_setup_trainer.py (hunk 6 above)
Comment (Ttopiac) [reply oai_agents/common/multi_setup_trainer.py]: I think 8 seeds is enough for our project and I will throw an error message to ask them to modify seeds setting in which function. Thank you both for suggestions.

File: oai_agents/common/population.py (hunk 7 above)
Comment (Ttopiac) [reply oai_agents/common/population.py]: I rename it to get_performance_based_population_by_layouts 
This name may be too long, but it also provides the most description and is also pretty clear what this function really work for.

File: oai_agents/agents/rl.py (hunk 12 above)
Comment (Ttopiac) [reply oai_agents/agents/rl.py:278]: We need to merge this branch and so I leave it to other pull request, but added a trello ticket for this sugggested changes.

File: oai_agents/common/population.py [hunk 20]
```diff
@@ -89,20 +89,20 @@ def ensure_enough_SP_agents(teammates_len,
     for eval_type in eval_types:
         if eval_type in TeamType.ALL_TYPES_BESIDES_SP:
             eval_agents_len += teammates_len
-        elif eval_type == TeamType.SELF_PLAY or eval_type == TeamType.SELF_PLAY_ADVERSARY:
-            eval_agents_len += 0
+        elif train_type in [TeamType.SELF_PLAY, TeamType.SELF_PLAY_ADVERSARY, TeamType.SELF_PLAY_STATIC_ADV, TeamType.SELF_PLAY_DYNAMIC_ADV]:
```
Comment (Jmiceli17) [inline oai_agents/common/population.py]: I think this should be `elif eval_type ...`

File: oai_agents/common/population.py [hunk 21]
```diff
@@ -89,20 +89,20 @@ def ensure_enough_SP_agents(teammates_len,
...
+            train_agents_len += 0
```
Comment (Jmiceli17) [inline oai_agents/common/population.py]: I think this should  be `eval_agents_len` right?
